File: object/object/src/Lec02/policy/discount/DefaultDiscountPolicy.java
package Lec02.policy.discount;

import Lec02.Money;
import Lec02.Screening;
import Lec02.condition.discount.discountCondition;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
public abstract class DefaultDiscountPolicy implements DiscountPolicy {

    private List<discountCondition> conditions = new ArrayList<>();

    public DefaultDiscountPolicy(discountCondition... conditions) {
        this.conditions = Arrays.asList(conditions);
    }

    @Override
    public Money calculateDiscountAmount(Screening screening) {
        for(discountCondition each : conditions) {
            if (each.isSatisfiedBy(screening)) {
                return getDiscountAmount(screening);
            }
        }

        return Money.ZERO;
    }

    abstract protected Money getDiscountAmount(Screening Screening);
}

File: object_Movie/src/java_to_kotlin/policy/discount/DefaultDiscountPolicy.kt
package java_to_kotlin.policy.discount

import java_to_kotlin.Money
import java_to_kotlin.Screening
import java_to_kotlin.condition.discount.DiscountCondition
import java.util.*

abstract class DefaultDiscountPolicy(vararg conditions: DiscountCondition?) : DiscountPolicy {
    private var conditions: MutableList<DiscountCondition> = ArrayList()

    init {
        this.conditions = Arrays.asList(*conditions)
    }

    override fun calculateDiscountAmount(screening: Screening): Money? {
        for (each in conditions) {
            if (each.isSatisfiedBy(screening)) {
                return getDiscountAmount(screening)
            }
        }
        return Money.Companion.ZERO
    }

    protected abstract fun getDiscountAmount(Screening: Screening): Money?
}
